### PR TITLE
`tensorflow`: Add members from `tensorflow.keras.metrics`

### PR DESCRIPTION
**Pull request by hoel-bagard**
These subs are taken from: https://github.com/hmc-cs-mdrissi/tensorflow_stubs/blob/main/stubs/tensorflow/keras/metrics.pyi

### CODE REVIEW

**Comment by hoel-bagard** [comment]
@hmc-cs-mdrissi If you have time to review it.

**Comment by JelleZijlstra** [comment]
FYI this has a merge conflict now.

**Comment by hoel-bagard** [comment]
@JelleZijlstra The CI was passing before merging with main, but now there is an error with the imports that I do not understand. Would you have an idea of what is causing it ?

**Comment by JelleZijlstra** [comment]
There was a failing stubtest job with no output; I reran it.

There is also a pytype crash:

```
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/pytd/visitors.py", line 512, in VisitNamedType
    raise KeyError(f"No {name} in module {module_name}") from e
KeyError: 'No ABCMeta in module importlib.resources.abc'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/runner/work/typeshed/typeshed/./tests/pytype_test.py", line 88, in run_pytype
    loader.finish_and_verify_ast(ast)
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 632, in finish_and_verify_ast
    self._modules[k].ast = self._resolve_external_types(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 526, in _resolve_external_types
    mod_ast = self._resolver.resolve_external_types(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 283, in resolve_external_types
    raise BadDependencyError(key, name) from e
pytype.load_pytd.BadDependencyError: No ABCMeta in module importlib.resources.abc, referenced from 'tensorflow.keras.metrics'
```

Which looks more mysterious, since I don't see how this PR relates to `importlib.resources`. cc @rchen152

File: stubs/tensorflow/tensorflow/keras/metrics.pyi [hunk 1]
```diff
@@ -1,9 +1,136 @@
...
+
+class Metric(tf.keras.layers.Layer[tf.Tensor, tf.Tensor], metaclass=ABCMeta):
+    def __init__(self, name: str | None = None, dtype: DTypeLike | None = None) -> None: ...
+    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...
+    def __call__(self, *args: Incomplete, **kwargs: Incomplete) -> Incomplete: ...
+    def __deepcopy__(self, memo: Incomplete | None = None) -> Incomplete: ...
+    @property
+    def _trackable_saved_model_saver(self) -> Incomplete: ...
+    @property
+    def trainable_weights(self) -> list[Incomplete]: ...
```
Comment (hmc-cs-mdrissi) [inline stubs/tensorflow/tensorflow/keras/metrics.pyi:24]: ? I think this method is inherited from Layer. Only methods defined new in class should be covered while inherited ones should be defined in parent.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi [hunk 2]
```diff
@@ -1,9 +1,136 @@
...
+
+_Output: TypeAlias = Tensor | dict[str, Tensor]
+
+class Metric(tf.keras.layers.Layer[tf.Tensor, tf.Tensor], metaclass=ABCMeta):
+    def __init__(self, name: str | None = None, dtype: DTypeLike | None = None) -> None: ...
+    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...
+    def __call__(self, *args: Incomplete, **kwargs: Incomplete) -> Incomplete: ...
+    def __deepcopy__(self, memo: Incomplete | None = None) -> Incomplete: ...
+    @property
+    def _trackable_saved_model_saver(self) -> Incomplete: ...
```
Comment (hmc-cs-mdrissi) [inline stubs/tensorflow/tensorflow/keras/metrics.pyi:22]: This method also looks like on tf.keras.layers.Layer.

It being private also makes it optional for you to include. Private stuff I mostly avoid unless I had code using it already.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi [hunk 3]
```diff
@@ -1,9 +1,136 @@
...
+from tensorflow.keras.layers import _Constraint
+from tensorflow.keras.regularizers import _Regularizer
+
+_Output: TypeAlias = Tensor | dict[str, Tensor]
+
+class Metric(tf.keras.layers.Layer[tf.Tensor, tf.Tensor], metaclass=ABCMeta):
+    def __init__(self, name: str | None = None, dtype: DTypeLike | None = None) -> None: ...
+    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...
+    def __call__(self, *args: Incomplete, **kwargs: Incomplete) -> Incomplete: ...
+    def __deepcopy__(self, memo: Incomplete | None = None) -> Incomplete: ...
```
Comment (hmc-cs-mdrissi) [inline stubs/tensorflow/tensorflow/keras/metrics.pyi:20]: Why `__deepcopy__`, `__new__`, and `__call__`? call should be inherited from layer class. deepcopy seems weird to define here. Unsure what special thing new is for either.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi [hunk 4]
```diff
@@ -1,9 +1,136 @@
...
+    def update_state(
+        self, y_true: TensorCompatible, y_pred: TensorCompatible, sample_weight: TensorCompatible | None = None
+    ) -> Operation | None: ...
+    @abstractmethod
+    def result(self) -> _Output: ...
+    @classmethod
+    def from_config(cls, config: dict[str, Any]) -> Self: ...
+    def get_config(self) -> dict[str, Any]: ...
+    @override
+    def add_weight(
```
Comment (hmc-cs-mdrissi) [inline stubs/tensorflow/tensorflow/keras/metrics.pyi:39]: This also looks like an inherited method so metrics class shouldn't have it defined.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi [hunk 5]
```diff
@@ -1,9 +1,136 @@
...
+    def non_trainable_weights(self) -> Incomplete: ...
+    def merge_state(self, metrics: Iterable[Self]) -> list[Operation]: ...
+    def reset_state(self) -> None: ...
+    @abstractmethod
+    def update_state(
+        self, y_true: TensorCompatible, y_pred: TensorCompatible, sample_weight: TensorCompatible | None = None
+    ) -> Operation | None: ...
+    @abstractmethod
+    def result(self) -> _Output: ...
+    @classmethod
```
Comment (hmc-cs-mdrissi) [inline stubs/tensorflow/tensorflow/keras/metrics.pyi:35]: Both from_config/get_config are also inherited from Layer.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 1 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Done in 8fd3fad1c

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 2 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Done in a6b5b10f5

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 3 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Removed in d45a599e9

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 4 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Done in e9246aa7b. It was already inherited from Layer indeed.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 5 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Done in eb6cef7f6

**Review by hmc-cs-mdrissi — APPROVED** [review]

**Comment by rchen152** [comment]
> There was a failing stubtest job with no output; I reran it.
> 
> There is also a pytype crash:
> 
> ```
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/pytd/visitors.py", line 512, in VisitNamedType
>     raise KeyError(f"No {name} in module {module_name}") from e
> KeyError: 'No ABCMeta in module importlib.resources.abc'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/runner/work/typeshed/typeshed/./tests/pytype_test.py", line 88, in run_pytype
>     loader.finish_and_verify_ast(ast)
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 632, in finish_and_verify_ast
>     self._modules[k].ast = self._resolve_external_types(
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 526, in _resolve_external_types
>     mod_ast = self._resolver.resolve_external_types(
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/site-packages/pytype/load_pytd.py", line 283, in resolve_external_types
>     raise BadDependencyError(key, name) from e
> pytype.load_pytd.BadDependencyError: No ABCMeta in module importlib.resources.abc, referenced from 'tensorflow.keras.metrics'
> ```
> 
> Which looks more mysterious, since I don't see how this PR relates to `importlib.resources`. cc @rchen152

Yeah, this was a weird one! Turns out it was an issue with how pytype resolves module aliases. I have a fix out for review, but I might not be able to release a new version of pytype until sometime next week, as we have a few big changes in the pipeline.

File: stubs/tensorflow/tensorflow/keras/metrics.pyi (hunk 4 above)
Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi:25]: @hmc-cs-mdrissi The `add_weight` from `keras.metrics.Metric` is different from the one from `keras.layers.Layer`, so I need to redefine it. However I'm facing the same error I had in another PR with `MultiHeadAttention`. Even if I add the method to the `stubtest_allowlist`, mypy still reports an error.

Comment (hmc-cs-mdrissi) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi:25]: `# type: ignore` works for that. stubtest allowlist is for runtime vs stub checks. Before that happens there's also a mypy pass. And you can use normal  `# type: ignore` to handle that.

Comment (hmc-cs-mdrissi) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi:25]: Generally get mypy passing first. No need to consider allowlist and then after mypy is green work on stubtest errors, and use allowlist when appropriate.

Comment (hoel-bagard) [reply stubs/tensorflow/tensorflow/keras/metrics.pyi]: Ok, thanks.

**Comment by rchen152** [comment]
> Yeah, this was a weird one! Turns out it was an issue with how pytype resolves module aliases. I have a fix out for review, but I might not be able to release a new version of pytype until sometime next week, as we have a few big changes in the pipeline.

Sorry, this took longer than expected, but pytype-2024.02.09 is out with this fix.

**Comment by hoel-bagard** [comment]
Thank you!

**Pull request merged**